### PR TITLE
Added Recipe

### PR DESCRIPTION
**Pull request by phuang00**
- setup recipes model
- support add and delete recipes

### CODE REVIEW

File: client/components/Recipe/RecipeComponent.vue [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    // Data from the stored recipe
+    recipe: {
+      type: Object,
+      required: true
+    }
+  },
+  data() {
+    return {
+      alerts: {} // Displays success/error messages encountered during recipe modification
+    };
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: Unnecessary, alerts are stored globally in state

File: client/components/Recipe/RecipeComponent.vue [hunk 2]
```diff
@@ -0,0 +1,88 @@
+<!-- Reusable component representing a single freet and its actions -->
+<!-- We've tagged some elements with classes; consider writing CSS using those classes to style them... -->
+
+<template>
+  <article class="recipe">
+    <section>
+      <button class="btn btn-primary m-2" @click="deleteRecipe">
+        Delete
+      </button>
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: Only show this button if user is the author of the recipe

File: client/components/Recipe/RecipeComponent.vue [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          const res = await r.json();
+          throw new Error(res.error);
+        }
+
+        this.$store.commit('refreshRecipes');
+
+        params.callback();
+      } catch (e) {
+        this.$set(this.alerts, e, 'error');
+        setTimeout(() => this.$delete(this.alerts, e), 3000);
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: Should be: 
```
this.$store.commit('alert', {
    message: e, 
    status: 'danger'
});
```

File: client/components/Recipe/RecipeComponent.vue [hunk 4]
```diff
@@ -0,0 +1,88 @@
+<!-- Reusable component representing a single freet and its actions -->
+<!-- We've tagged some elements with classes; consider writing CSS using those classes to style them... -->
+
+<template>
+  <article class="recipe">
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: Add this inside class: `border rounded my-2 p-4`

File: client/components/Recipe/RecipeComponent.vue [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+};
+</script>
+
+<style scoped>
+.recipe {
+  border: 1px solid #111;
+  padding: 20px;
+  position: relative;
+}
+</style>
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: Can be removed, see bootstrap styling comment above

File: client/components/Recipe/RecipeComponent.vue [hunk 6]
```diff
@@ -0,0 +1,88 @@
+<!-- Reusable component representing a single freet and its actions -->
+<!-- We've tagged some elements with classes; consider writing CSS using those classes to style them... -->
+
+<template>
+  <article class="recipe">
+    <section>
+      <button class="btn btn-primary m-2" @click="deleteRecipe">
```
Comment (jlin00) [inline client/components/Recipe/RecipeComponent.vue]: change the class to "btn btn-danger btn-sm my-2 bi bi-trash"

File: client/components/common/BlockForm.vue [hunk 7]
```diff
@@ -251,6 +251,20 @@
             </label>
           </div>
         </div>
+        <div v-else-if="field.type === 'cookTime'" class="row">
+          <div class="col-9">
```
Comment (jlin00) [inline client/components/common/BlockForm.vue]: change class to col-9

File: client/components/common/BlockForm.vue [hunk 8]
```diff
@@ -251,6 +251,20 @@
...
+            <input 
+              class="form-control"
+              type="number" 
+              :name="field.id" 
+              :value="field.value"
+              :placeholder="field.placeholder"
```
Comment (jlin00) [inline client/components/common/BlockForm.vue]: Add an informative placeholder for this generic. Something like
`placeholder="Enter number..."`

File: client/components/Recipe/AddRecipeForm.vue [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  data() {
+    return {
+      url: '/api/recipes',
+      method: 'POST',
+      hasBody: true,
+      fields: [
+        { type: 'text', id: 'name', label: 'Name', value: '' }, 
+        { type: 'ingredients', id: 'ingredients', label: 'Ingredients', name: '', quantity: '', unit: '', ingredients: [] },
+        { type: 'text', id: 'instructions', label: 'Instructions', value: '' }, 
+        { type: 'cookTime', id: 'cookTime', label: 'Cook Time', value: '' },
```
Comment (jlin00) [inline client/components/Recipe/AddRecipeForm.vue]: Change label to 'Cook Time (in minutes)'

File: server/recipe/collection.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * Get all the recipes that contains a list of ingredients
+   *
+   * @param {Array<string>} ingredients - The list of ingredient names that the recipe contains.
+   * @return {Promise<HydratedDocument<Recipe>[]>} - An array of all of the recipes with these ingredients sorted by name alphabetically.
+   */
+  static async findAllByIngredients(ingredients: Array<string>): Promise<Array<HydratedDocument<Recipe>>> {
+    return RecipeModel.find({ingredients: {$all: ingredients.map(name => ({$elemMatch: {'name': name}}))}}).sort('name').populate('author').populate('ingredients');
+  }
```
Comment (jlin00) [inline server/recipe/collection.ts]: For reference, here is the code I used to find freets matching a list of authors/tags:
```
return FreetModel.find({
      $or: [
        {authorId: {$in: authors}},
        {tags: {$in: tags}}
      ]
    }).sort({dateModified: -1}).populate(['authorId', 'tags', 'upvotes', 'downvotes', 'links']);
```

File: server/recipe/collection.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * Get all the recipes with keywords in name
+   *
+   * @param {string} keywords - The keywords to search for
+   * @return {Promise<HydratedDocument<Recipe>[]>} - An array of all of the recipes with the keywords sorted by name alphabetically.
+   */
+   static async findAllByKeywords(keywords: string): Promise<Array<HydratedDocument<Recipe>>> {
```
Comment (jlin00) [inline server/recipe/collection.ts]: rename from keywords to keyword, since it is just one

File: server/recipe/collection.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  /**
+   * Get all the recipes with keywords in name
+   *
+   * @param {string} keywords - The keywords to search for
+   * @param {Array<string>} ingredients - The list of ingredient names that the recipe contains.
+   * @return {Promise<HydratedDocument<Recipe>[]>} - An array of all of the recipes with the keywords and ingredients sorted by name alphabetically.
+   */
+   static async findAllByKeywordsAndIngredients(keywords: string, ingredients: Array<string>): Promise<Array<HydratedDocument<Recipe>>> {
```
Comment (jlin00) [inline server/recipe/collection.ts]: rename from keywords to keyword

File: server/recipe/collection.ts [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    return RecipeModel.find({name: { $regex: keywords, $options: "i" }, ingredients: {$all: ingredients.map(name => ({$elemMatch: {'name': name}}))}}).sort('name').populate('author').populate('ingredients');
```
Comment (jlin00) [inline server/recipe/collection.ts]: See previous comments for filtering recipes by matching ingredients

File: server/recipe/collection.ts [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  /**
+   * Update a recipe with the new information
+   *
+   * @param {Types.ObjectId | string} receipeId - The id of the grocery to be updated
```
Comment (jlin00) [inline server/recipe/collection.ts]: Do a find all and replace receipe with recipe (typo)

File: server/recipe/router.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+ * @return {string} - A success message
+ * @throws {403} - If the user is not logged in or if the user is trying to delete someone else's recipe
+ * @throws {404} - If the recipeId is not valid
+ */
+router.delete(
+  '/:recipeId?',
+  [
+    userValidator.isUserLoggedIn,
+    recipeValidator.isValidRecipeModifier,
+    recipeValidator.isRecipeExists,
```
Comment (jlin00) [inline server/recipe/router.ts:77]: `isRecipeExists` check has to come before `isValidRecipeModifier`

**Review by jlin00 — CHANGES_REQUESTED** [review]
Minor bug fixes and cosmetic changes. Should be good to merge after requested changes are submitted.

File: client/components/Recipe/RecipeComponent.vue (hunk 1 above)
Comment (phuang00) [reply client/components/Recipe/RecipeComponent.vue]: deleted

File: client/components/Recipe/RecipeComponent.vue (hunk 2 above)
Comment (phuang00) [reply client/components/Recipe/RecipeComponent.vue]: updated

File: client/components/Recipe/RecipeComponent.vue (hunk 3 above)
Comment (phuang00) [reply client/components/Recipe/RecipeComponent.vue]: updated

File: client/components/Recipe/RecipeComponent.vue (hunk 4 above)
Comment (phuang00) [reply client/components/Recipe/RecipeComponent.vue]: updated

File: client/components/Recipe/RecipeComponent.vue (hunk 5 above)
Comment (phuang00) [reply client/components/Recipe/RecipeComponent.vue]: deleted

File: server/recipe/collection.ts (hunk 11 above)
Comment (phuang00) [reply server/recipe/collection.ts]: updated

File: server/recipe/collection.ts (hunk 12 above)
Comment (phuang00) [reply server/recipe/collection.ts]: resolved

File: client/components/common/BlockForm.vue (hunk 7 above)
Comment (phuang00) [reply client/components/common/BlockForm.vue]: do u mean remove col-9?

**Pull request merged**